Implement "View World"

vincent-grosbois opened the pull request:
Here's a big PR that enables the "View World" window
What is missing:
- incorrect offset for some icons
- resource name is not written in the icons

I did some timings:
- pre-generation of the terrain takes 0.1 sec at map loading for XL maps
- adding fog etc for the terrain takes 0.03 sec when loading the view world screen for XL map
- redrawing the map when scrolling etc takes a few ms

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,653 @@
+/***************************************************************************
```
Please add file footer according to [this page](https://github.com/ihhub/fheroes2/wiki/Copyright-standard).

ihhub commented on src/fheroes2/kingdom/view_world.h:
```diff
@@ -0,0 +1,120 @@
+/***************************************************************************
```
Please add file footer according to [this page](https://github.com/ihhub/fheroes2/wiki/Copyright-standard).

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,653 @@
+                        const int dstx0 = xx * TileSizePerZoomLevel[0];
+                        const int dstx1 = xx * TileSizePerZoomLevel[1];
+                        const int dstx2 = xx * TileSizePerZoomLevel[2];
+                        fheroes2::Blit( blackFogTile0, zoomLevel0, dstx0, dsty0 );
+                        fheroes2::Blit( blackFogTile1, zoomLevel1, dstx1, dsty1 );
+                        fheroes2::Blit( blackFogTile2, zoomLevel2, dstx2, dsty2 );
```
Please use `fheroes2::Fill` function instead so you won't need any extra images for this.

ihhub commented on src/fheroes2/maps/maps_tiles.h:
```diff
@@ -294,9 +294,10 @@ namespace Maps
         friend TiXmlElement & operator>>( TiXmlElement &, Tiles & );
 #endif
 
+    public:
```
We strictly do no allow any members to be public.

ihhub commented on src/fheroes2/world/world.h:
```diff
@@ -321,6 +322,9 @@ class World : protected Size
     std::vector<MapRegion> _regions;
     PlayerWorldPathfinder _pathfinder;
+
+public:
```
We strictly do no allow any members to be public.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,653 @@
+
+    Cursor & cursor = Cursor::Get();
+    const int oldcursor = cursor.Themes();
+    cursor.Hide();
```
Please remove this and line 559 as they are redundant for the new redering engine.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,653 @@
+    }
+}
+
+void drawAddon( fheroes2::Image & dst, const Maps::Addons & addon, int dstx, int dsty, int quantity2, uint32_t UIDToIgnore )
```
This and the function below are alsmot the same to `void Maps::Tiles::RedrawAddon` and `void Maps::Tiles::RedrawObjects` methods. Why don't we modify `Interface::GameArea` ROI like it's done for puzzle for the moment of generating image? Then we can call `void Interface::GameArea::Redraw` method with needed parameters which will generate for us needed image (maybe we have to add some small checks like for monster drawings) but this approach will be much cleaner and smaller. You could take a look into `fheroes2::Image Interface::GameArea::GenerateUltimateArtifactAreaSurface` method as a reference. After getting needed image we just need to draw extra objects like heroes or monsters.
What do you think?

ihhub reviewed (changes requested):
Hi @vincent-grosbois , good work! I proposed another approach to do the same. Could you please take a look into given comments?

ihhub commented:
Hi @vincent-grosbois could you please rebase your branch from master? I'm not able to compile it on my machine once I put your changes on the top of the master branch.

vincent-grosbois commented:
> 
> 
> Hi @vincent-grosbois could you please rebase your branch from master? I'm not able to compile it on my machine once I put your changes on the top of the master branch.

it should be good now

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+        Interface::GameArea gamearea = Interface::Basic::Get().GetGameArea();
+
+        // need a copy so initial full terrain image can be reused
+        cachedImages[3] = world.getViewWorldCache().GetCachedImage();
```
Storing cached image in world might be not a good idea since we are no releasing resources after exiting World View mode. What if we calculate it every time inside this constructor? How slow it is to do such?

vincent-grosbois replied:
I think it's 0.1 sec at map loading for the biggest map

ihhub replied:
The timing really depends on hardware. What are specs of your machine?

vincent-grosbois replied:
i7-6700 HQ@2.6Ghz
8GB of ram
so not super recent laptop

ihhub replied:
Hmm... then it should be okay to do calculations upon clicking View World button.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+void ViewWorldTerrainCache::Recompute()
+{
+    // Just draw the tiles layer
+    _cachedTerrain = fheroes2::Image( world.w() * TILEWIDTH, world.h() * TILEWIDTH );
```
It's a huge image after all: 2 * 192 * 32 * 192 * 32 = 72 MB. We can't do this as some systems such as PlayStation Vita doesn't have so much RAM available for our application.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+{
+    // Just draw the tiles layer
+    _cachedTerrain = fheroes2::Image( world.w() * TILEWIDTH, world.h() * TILEWIDTH );
+    Interface::GameArea gamearea = Interface::Basic::Get().GetGameArea();
```
We do we make a copy of game are here? Can't we remember the original roi, do computations and then return roi back?

vincent-grosbois replied:
Yes but whats is the reason for doing it that way ? It's logically a different GameArea that I want to draw on, so I don't see the upsides of doing weird temporary change of values. I think that GameArea class is very light too

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+    fheroes2::Image cachedImages[4]; // One image per zoom Level
+
+    // Compute complete world map, and save it for all zoom levels
+    CacheForMapWithResources( bool revealAll )
```
Please make this constructor `explicit` and input argument `const`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+
+struct CacheForMapWithResources
+{
+    fheroes2::Image cachedImages[4]; // One image per zoom Level
```
Here's a problem with memory: we store 72 MB image in world map, then we have 1 image as a copy of it, and 3 more images of 18 MB, 4.5 MB and 1.125 MB. Overall it's 167.625 MB which is an overkill for memory consumption.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+            world.getViewWorldCache().Recompute();
+        }
+
+        Interface::GameArea gamearea = Interface::Basic::Get().GetGameArea();
```
Is there a reason why we make a copy of game area?

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+    return result;
+}
+
+int colorToOffsetICN( int color )
```
Since this is a static function within the source file please create an unnamed namespace just after header inclusions and put this function here. This is to reduce the scope of function and to improve readability of the code as we can know what extra we have in the file.
Also please make input argument as `const`.

ihhub commented on src/fheroes2/kingdom/view_world.h:
```diff
@@ -0,0 +1,104 @@
+
+        ZoomLevel _zoomLevel;
+        Point _center;
+        Rect _roiForZoomLevels[4];
```
Please prefer **fheroes2::Rect** over **Rect** as the old class is limited to int16_t and causes conversion warnings in the code.

ihhub commented on src/fheroes2/kingdom/view_world.h:
```diff
@@ -0,0 +1,104 @@
+        Rect GetROIinTiles() const;
+
+        ZoomLevel _zoomLevel;
+        Point _center;
```
Please prefer **fheroes2::Point** over **Point** as the old class is limited to int16_t and causes conversion warnings in the code.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+        display.render();
+    }
+
+    cursor.Hide();
```
Please remove `cursor.Hide()` and `cursor.Show()` as it's a legacy code and make no difference for us.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+                            display.height() - viewWorldSprite.height() - BORDERWIDTH );
+        }
+
+        cursor.Show();
```
Please remove this line as it does nothing here.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+        else if ( le.MousePressLeft( visibleScreenInPixels ) ) {
+            if ( isDrag ) {
+                Point newMousePos = le.GetMouseCursor();
+                Point newRoiCenter;
+                newRoiCenter.x = initRoiCenter.x - ( newMousePos.x - initMousePos.x ) * TILEWIDTH / TileSizePerZoomLevel[currentROI._zoomLevel];
+                newRoiCenter.y = initRoiCenter.y - ( newMousePos.y - initMousePos.y ) * TILEWIDTH / TileSizePerZoomLevel[currentROI._zoomLevel];
```
Please use constructor to assign values in place.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+    fheroes2::Blit( viewWorldSprite, display, display.width() - viewWorldSprite.width() - BORDERWIDTH, display.height() - viewWorldSprite.height() - BORDERWIDTH );
+
+    // Zoom button
+    Point dst_pt2( display.width() - RADARWIDTH + 16, display.height() - 177 );
```
Please use `fheroes2::Point` type instead.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+    buttonZoom.draw();
+
+    // Exit button
+    Point dst_pt( display.width() - RADARWIDTH + 16, display.height() - 69 );
```
Please use `fheroes2::Point` type instead.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+
+    fheroes2::Display & display = fheroes2::Display::instance();
+
+    const s32 world_w = world.w();
```
Please use `int32_t` instead of `s32`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+
+//#define VIEWWORLD_DEBUG_ZOOM_LEVEL  // Activate this when you want to debug this window. It will provide an extra zoom level at 1:1 scale
+
+static const int TileSizePerZoomLevel[4] = {4, 6, 12, 32};
```
Please put these variables as well as other static functions declared in this file under unnamed namespace. Variables don't need `static` keyword in this case. It's good to know what helper code exist in this source code.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+
+//#define VIEWWORLD_DEBUG_ZOOM_LEVEL  // Activate this when you want to debug this window. It will provide an extra zoom level at 1:1 scale
+
+static const int TileSizePerZoomLevel[4] = {4, 6, 12, 32};
```
We might not need to cache images. Please use `ICN::GROUND12`, `ICN::GROUND6` and `ICN::GROUND4` for terrain generation. It should be much more efficient.

ihhub replied:
Also I think these ICNs might hel you with implementation:
- ICN::LETTER12, ICN::LETTER6, ICN::LETTER4
- ICN::MISC12, ICN::MISC6, ICN::MISC4
- ICN::VWFLAG12, ICN::VWFLAG6, ICN::VWFLAG4

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+    const Size & sizeInPixels = Interface::Basic::Get().GetGameArea().GetROI();
+
+    // how many pixels from "world map" we can see in "view world" window, given current zoom
+    const int pixelsW = sizeInPixels.w * TILEWIDTH / TileSizePerZoomLevel[(int)zoomLevel];
```
Please always use C++ casting over C-casting.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,507 @@
+    int offsetPixelsY = TileSizePerZoomLevel[ROI._zoomLevel] * ROI.GetROIinPixels().y / TILEWIDTH;
+
+    // Crop the cache image based on the screen size
+    fheroes2::Sprite s = fheroes2::Crop( image, offsetPixelsX, offsetPixelsY, ROIScreen.w, ROIScreen.h );
```
Please make this variable `const` and call it with some meaningful name like `cropped`.

ihhub reviewed (changes requested):
Hi @vincent-grosbois , I left multiple comments here as a part of first review. Could you please take a look?

vincent-grosbois commented:
@ihhub 
I've addressed some comments, but it's not finished yet.
I don't know what you mean about using ICN::GROUND16 etc. Am I supposed to change the whole part about drawing world and just drawing these ICN instead ? How do I know which index to select?

ihhub commented:
> @ihhub
> I've addressed some comments, but it's not finished yet.
> I don't know what you mean about using ICN::GROUND16 etc. Am I supposed to change the whole part about drawing world and just drawing these ICN instead ? How do I know which index to select?

Hi @vincent-grosbois ,
GROUND are tiles for zoomed images. In other words they're resized versions of original 32x32 tiles. MISC is for objects on map and LETTER is a letter which should be used on these objects. To have a clear picture please do this:
- add `#include "image_tool.h"` into agg.cpp
- inside `bool LoadModifiedICN( int id )` function add this code:
```
            case ICN::GROUND12:
                LoadOriginalICN( id );
                for ( size_t i = 0; i < _icnVsSprite[id].size(); ++i ) {
                    Save( _icnVsSprite[id][i], "your_path" + std::to_string(i) + ".bmp");
                }
```

so that youcan see what these ICNs are.

vincent-grosbois commented:
@ihhub I've updated the PR, normally it correctly uses the new resize API

ihhub commented:
Hi @vincent-grosbois , few things to notice:
- my FPS goes crazy in this mode. A wooping 400 FPS :) Are we doing everything in a loop?
- there's a black section on the right of the interface. I recommend to put View World there and just draw the rest of background

![image](https://user-images.githubusercontent.com/19829520/106469126-d76f2b80-64d9-11eb-8809-6ade3f3a6b3a.png)

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+        }
+
+        cursor.Show();
+        display.render();
```
We should put this line in the block before as there's no reason to redraw if nothing's changed.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+            return -1;
+        }
+    }
+
```
Please remove this empty line.

ihhub commented on src/fheroes2/maps/maps_tiles.cpp:
```diff
@@ -2624,15 +2615,25 @@ void Maps::Tiles::RedrawFogs( fheroes2::Image & dst, int color ) const
     }
 }
 
+const Maps::Addons & Maps::Tiles::getAddonsLevel1() const
+{
+    return addons_level1;
+}
+
+const Maps::Addons & Maps::Tiles::getAddonsLevel2() const
+{
+    return addons_level2;
+}
+
```
These methods are not used anywhere. Should we remove them?

ihhub commented on src/fheroes2/maps/maps_tiles.h:
```diff
@@ -300,7 +308,6 @@ namespace Maps
 
         Addons addons_level1;
         Addons addons_level2; // 16
-
```
Please revert this line. Let's keep minimal changes for this pull request.

ihhub commented on src/fheroes2/gui/interface_radar.cpp:
```diff
@@ -98,17 +99,25 @@ uint8_t GetPaletteIndexFromColor( int color )
     default:
         break;
     }
-
```
Please revert this line. Let's keep minimal changes for this pull request.

vincent-grosbois commented:
> Are we doing everything in a loop?
What do you mean?
>     * there's a black section on the right of the interface. I recommend to put View World there and just draw the rest of background
I'll see what I can do, it's a good idea to show the rest of the map indeed. But I don't know how I can make it so the borders on the right don't look ugly. Also I'll have to check all resolutions

ihhub commented on src/fheroes2/gui/interface_gamearea.cpp:
```diff
@@ -149,6 +149,9 @@ void Interface::GameArea::Redraw( fheroes2::Image & dst, int flag, bool isPuzzle
     int32_t maxY = tileROI.y + tileROI.h;
 
+    // ground and bottom layer
+    const bool drawBottom = ( flag & LEVEL_BOTTOM ) == LEVEL_BOTTOM;
```
Please move this back to the original place since there're no need move this variable.

ihhub commented on src/fheroes2/gui/interface_gamearea.cpp:
```diff
@@ -262,13 +263,13 @@ void Interface::GameArea::Redraw( fheroes2::Image & dst, int flag, bool isPuzzle
     const Heroes * hero = drawHeroes ? GetFocusHeroes() : NULL;
 
-    if ( hero && hero->GetPath().isShow() ) {
+    if ( hero && hero->GetPath().isShow() && ( flag & LEVEL_ROUTES ) ) {
```
Please declare an extra variable `const bool darwRoutes = ( flag & LEVEL_ROUTES ) != 0;` and use it here.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+    explicit CacheForMapWithResources( const bool revealAll )
+    {
+#ifdef VIEWWORLD_DEBUG_ZOOM_LEVEL
+        const int nCachedLevels = 4;
+#else
+        const int nCachedLevels = 3;
```
Please do **not** use Hungarian variable naming, just call it as `cachedLevels`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+        assert( ( world.h() * TILEWIDTH ) % blockSizeY == 0 );
+
+        // Create temporary image where we will draw blocks of the main map on
+        fheroes2::Image temporaryImg = fheroes2::Image( blockSizeX, blockSizeY );
```
Please use normal constructor `fheroes2::Image temporaryImg( blockSizeX, blockSizeY );` as it's the fastest way.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+        const int32_t blockSizeY = TILEWIDTH * 18;
+
+        // Assert will fail in case we add non-standard map sizes, otherwise standard map sizes are multiples of 18 tiles
+        assert( ( world.w() * TILEWIDTH ) % blockSizeX == 0 );
+        assert( ( world.h() * TILEWIDTH ) % blockSizeY == 0 );
```
Please cache `world.w() * TILEWIDTH` and `world.h() * TILEWIDTH` and use them in the loops below and here.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+                gamearea.Redraw( temporaryImg, drawingFlags );
+
+                for ( int i = 0; i < nCachedLevels; ++i ) {
+                    const int blockSizeResizedX = blockSizeX * TileSizePerZoomLevel[i] / TILEWIDTH;
```
Please cache `TileSizePerZoomLevel[i] / TILEWIDTH`.

vincent-grosbois replied:
I'll do it but I really don't see the point, makes things less readable IMO
In term of code generated I'm sure it's not changing anything

vincent-grosbois replied:
I can't cache it actually, the whole point is that this division does not yield an integer, so we need to do the full computation 
> blockSizeX * TileSizePerZoomLevel[i] / TILEWIDTH
and
> x * TileSizePerZoomLevel[i] / TILEWIDTH
to have exact result

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+void DrawWorld( const ViewWorld::ZoomROIs & ROI, CacheForMapWithResources & cache )
+{
+    fheroes2::Display & display = fheroes2::Display::instance();
+    display.fill( 0 );
```
Is there a strong need to fill the whole image? I think what we need is to update certain parts of the display.

vincent-grosbois replied:
I'm not sure what you mean
There are certains parts of the display that are potentially never drawn over on the next frame, like the  part of the frame where no map is being drawn. That's why I set everything to 0 first

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+    int offsetPixelsY = TileSizePerZoomLevel[static_cast<int>( ROI._zoomLevel )] * ROI.GetROIinPixels().y / TILEWIDTH;
+
+    // Crop the cache image based on the screen size
+    const fheroes2::Sprite cropped = fheroes2::Crop( image, offsetPixelsX, offsetPixelsY, ROIScreen.w, ROIScreen.h );
+
+    fheroes2::Blit( cropped, display, BORDERWIDTH + ( offsetPixelsX < 0 ? -offsetPixelsX : 0 ), BORDERWIDTH + ( offsetPixelsY < 0 ? -offsetPixelsY : 0 ) );
```
Just use `fheroes2::Blit` call directly without any cropping. This function has an overloaded version to support custom width and height and it does internal 'cropping' inside without an extra image.

vincent-grosbois replied:
I tried but it seems it doesn't work well.
I think in some situations (if offsetPixelsX  > 0), we fail the Verify() check in Blit() and we do'nt draw anything

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+    const int offsetX = roiPixels.x * tileSize / TILEWIDTH;
+    const int offsetY = roiPixels.y * tileSize / TILEWIDTH;
+
+    Rect roiTiles = ROI.GetROIinTiles();
```
Please make this variable `const`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+    const int32_t world_w = world.w();
+    const int32_t world_h = world.h();
+
+    Rect roiPixels = ROI.GetROIinPixels();
```
Please make this variable `const`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+    const int32_t maxTileY = clamp( roiTiles.y + roiTiles.h + 2, 0, world_h );
+
+    for ( int32_t yy = minTileY; yy < maxTileY; yy += 1 ) {
+        const int dsty = yy * tileSize - offsetY + BORDERWIDTH;
```
Please call it as `posY` or `dstY`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+        const int dsty = yy * tileSize - offsetY + BORDERWIDTH;
+
+        for ( int32_t xx = minTileX; xx < maxTileX; xx += 1 ) {
+            const int dstx = xx * tileSize - offsetX + BORDERWIDTH;
```
Please call it as `posX` or `dstX`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+    const int32_t minTileY = clamp( roiTiles.y - 2, 0, world_h );
+    const int32_t maxTileY = clamp( roiTiles.y + roiTiles.h + 2, 0, world_h );
+
+    for ( int32_t yy = minTileY; yy < maxTileY; yy += 1 ) {
```
Please use `y` instead of `yy`.

ihhub replied:
Also please use normal pre-increment of the variable instead of `+= 1`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+    for ( int32_t yy = minTileY; yy < maxTileY; yy += 1 ) {
+        const int dsty = yy * tileSize - offsetY + BORDERWIDTH;
+
+        for ( int32_t xx = minTileX; xx < maxTileX; xx += 1 ) {
```
Please use `x` instead of `xx`.

ihhub replied:
Also please use normal pre-increment of the variable instead of `+= 1`.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+
+namespace
+{
+    const int TileSizePerZoomLevel[4] = {4, 6, 12, 32};
+    const int IcnPerZoomLevel[4] = {ICN::MISC4, ICN::MISC6, ICN::MISC12, ICN::MISC12};
+    const int IcnPerZoomLevelFlags[4] = {ICN::VWFLAG4, ICN::VWFLAG6, ICN::VWFLAG12, ICN::VWFLAG12};
```
Please name all variable from non-capital letter.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+                        case MP2::OBJ_MINES:
+                        case MP2::OBJ_SAWMILL:
+                            if ( revealMines || !tile.isFog( color ) ) { // draw mine now, hero on top after the switch
+                                int colorOffset = colorToOffsetICN( tile.QuantityColor() );
```
Here we shadow the previous variable. Please use the previous variable or rename this one.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,511 @@
+    // Creates fixed radar on top-right, even if hidden interface
+    Interface::Radar radar = Interface::Radar::MakeRadarViewWorld( interface.GetRadar() );
+
+    Rect worldMapROI = interface.GetGameArea().GetVisibleTileROI();
+    Rect visibleScreenInPixels = interface.GetGameArea().GetROI();
```
Please make these variables `const`.

ihhub commented on src/fheroes2/world/world.h:
```diff
@@ -31,6 +31,7 @@
 #include "maps.h"
 #include "maps_objects.h"
 #include "maps_tiles.h"
+#include "view_world.h"
```
Do we need this file inclusion?

ihhub reviewed (changes requested):
Hi @vincent-grosbois , I left multiple comments here. Could you please take a look?

ihhub commented:
> What do you mean?
I pointed a place to fix this :)

Also I don't see that in the original game there are any changes related to Radar. The only thing changes is visible area. Could you please check it? If it's true then we don't need all changes for radar but a simple set of visible are for it.

vincent-grosbois commented:
@ihhub I adressed all comments in this new commit, I think

> Also I don't see that in the original game there are any changes related to Radar. The only thing changes is visible area. Could you please check it? If it's true then we don't need all changes for radar but a simple set of visible are for it.
We need to change the radar because it has to handle changing area size (for the rectangle) and it also has to handle the change of zoom level

ihhub commented on src/fheroes2/kingdom/view_world.h:
```diff
@@ -0,0 +1,75 @@
+
+enum class ViewWorldMode : int
+{
+    OnlyVisible = 0, // Only show what is currently under fog of war
```
Should the comment state the opposite: `Only show what is current not under fog of war`?

ihhub reviewed (changes requested):
Hi @vincent-grosbois , I left multiple comments here. Could you please take a look?

vincent-grosbois commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,515 @@
+    const Rect visibleScreenInPixels = interface.GetGameArea().GetROI();
+
+    // Initial view is centered on where the player is centered
+    fheroes2::Point viewCenterInPixels( worldMapROI.x * TILEWIDTH + visibleScreenInPixels.w / 2, worldMapROI.y * TILEWIDTH + visibleScreenInPixels.h / 2 );
```
not possible due to the if check afterwards

vincent-grosbois commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,515 @@
+    const int offsetPixelsY = tileSizePerZoomLevel[static_cast<int>( ROI._zoomLevel )] * ROI.GetROIinPixels().y / TILEWIDTH;
+
+    // Crop the cache image based on the screen size
+    const fheroes2::Sprite cropped = fheroes2::Crop( image, offsetPixelsX, offsetPixelsY, ROIScreen.w + 2 * BORDERWIDTH + RADARWIDTH, ROIScreen.h );
```
Hello
I've tried to replace this with:
    ```
    fheroes2::Point inPos( offsetPixelsX, offsetPixelsY );
        fheroes2::Point outPos( BORDERWIDTH + ( offsetPixelsX < 0 ? -offsetPixelsX : 0 ), BORDERWIDTH + ( offsetPixelsY < 0 ? -offsetPixelsY : 0 ) );
        fheroes2::Size outSize( ROIScreen.w + 2 * BORDERWIDTH + RADARWIDTH, ROIScreen.h );
        fheroes2::Rect outputROI( outPos , outSize);

        fheroes2::FitToRoi( image, inPos, display, outPos, outSize, outputROI );

        fheroes2::Blit( image, inPos, display, outPos, outSize );

But basically it still fails. Basically when we passe inX < 0 or inY < 0 we abort drawing. It seems like a bug to me
```

vincent-grosbois replied:
I think I found a way to do it

vincent-grosbois commented:
hi @ihhub I think I adressed all the issues

ihhub commented on src/fheroes2/gui/interface_border.h:
```diff
@@ -28,7 +28,7 @@
 namespace Interface
 {
-    void GameBorderRedraw( void );
+    void GameBorderRedraw( const bool viewWorld );
```
To be more understandable I'm proposing to name this variable as `viewWorldMode`.

ihhub commented on src/fheroes2/gui/interface_radar.cpp:
```diff
@@ -305,17 +326,125 @@ void Interface::Radar::RedrawObjects( int color )
+    else
+        sw = GetChunkSize( areah, world_h );
+
+    fheroes2::Image sf( sw, sw );
```
This image should go away.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,531 @@
+        // Compute complete world map, and save it for all zoom levels
+        explicit CacheForMapWithResources( const bool revealAll )
+        {
+#ifdef VIEWWORLD_DEBUG_ZOOM_LEVEL
+            const int cachedLevels = 4;
+#else
+            const int cachedLevels = 3;
+#endif
```
Shall we use `std::vector<fheroes2::Image> cachedImages` instead? Then here we can do like this:
```
#ifdef VIEWWORLD_DEBUG_ZOOM_LEVEL
            cachedImages.resize( 4 );
#else
            cachedImages.resize( 3 );
#endif
```
and lower in this code use `cachedImages.size()`. What do you think?

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,531 @@
+            LocalEvent::Get().PauseCycling();
+
+            for ( int i = 0; i < cachedLevels; ++i ) {
+                cachedImages[i] = fheroes2::Image( world.w() * tileSizePerZoomLevel[i], world.h() * tileSizePerZoomLevel[i] );
```
Please use `resize()` call instead.

ihhub commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,531 @@
+        fheroes2::Image cachedImages[4]; // One image per zoom Level
+
+        // Compute complete world map, and save it for all zoom levels
+        explicit CacheForMapWithResources( const bool revealAll )
```
If we don't expect default constructor to be called please delete it: `CacheForMapWithResources() = delete;`.

ihhub commented on src/fheroes2/kingdom/view_world.h:
```diff
@@ -0,0 +1,75 @@
+    {
+        ZoomROIs( const ZoomLevel zoomLevel, const fheroes2::Point & centerInPixels );
+
+        bool ChangeZoom( const bool zoomIn, const bool cyle = false );
```
`cycle` is misspelled.

ihhub reviewed (changes requested):
Hi @vincent-grosbois , good job! We're almost there. Could you please rebase your branch from master and check the comments?

vincent-grosbois commented:
hi @ihhub , here's a new review that includes all changes

ihhub reviewed (changes requested):
Hi @vincent-grosbois , I left multiple comments here. Could you please take a look?

vincent-grosbois commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,529 @@
+
+bool ViewWorld::ZoomROIs::ChangeCenter( const fheroes2::Point & centerInPixels )
+{
+    fheroes2::Point newCenter = centerInPixels;
+    newCenter.x = clamp( newCenter.x, 0, world.w() * TILEWIDTH );
+    newCenter.y = clamp( newCenter.y, 0, world.h() * TILEWIDTH );
```
you're aware that it's going to produce the same assembly right ?

vincent-grosbois commented on src/fheroes2/kingdom/view_world.cpp:
```diff
@@ -0,0 +1,529 @@
+    // Creates fixed radar on top-right, even if hidden interface
+    Interface::Radar radar = Interface::Radar::MakeRadarViewWorld( interface.GetRadar() );
+
+    const Rect worldMapROI = interface.GetGameArea().GetVisibleTileROI();
+    const Rect visibleScreenInPixels = interface.GetGameArea().GetROI();
```
I dcan't do it for GetVisibleTileROI as it's returning a new object

vincent-grosbois commented:
@ihhub  done

ihhub commented on src/fheroes2/gui/interface_radar.cpp:
```diff
@@ -305,17 +326,122 @@ void Interface::Radar::RedrawObjects( int color ) const
-/* redraw radar cursor */
-void Interface::Radar::RedrawCursor( void )
+// Redrawing method for "View World".
+void Interface::Radar::RedrawObjects( int color, ViewWorldMode flags ) const
```
Let's combine this and the original `RedrawObjects` methods. The only difference is presence of flags. To compensate an extra computations for normal radar let's do this:
- move `GetChunkSize` function at the top of the file under unnamed namespace and modify it's body to:
```
int GetChunkSize( int size1, int size2 )
{
    if ( size1 > size2 ) {
        const int res = size1 / size2;
        if ( ( size1 % size2 ) * 10 > size2 ) {
            return res + 1;
        }
        return res;
    }

    return 1;
}
```
this will be slightly faster than before.
- in `RedrawObjects` method inside the loop make these changes (notice swapped conditions):
```
#ifdef WITH_DEBUG
            const bool visibleTile = revealAll || IS_DEVEL() || !tile.isFog( color );
#else
            const bool visibleTile = revealAll || !tile.isFog( color );
#endif
```
- change all uint32_t to int32_t (we're introducing implicit type conversion)
- cache `rect.y + offset.y` and `rect.x + offset.x` before the loop

ihhub reviewed (changes requested):
Hi @vincent-grosbois , few more things are left to be addressed:
- code duplication removal in Interface::Radar. I described the way how to compensate performance drop
- we should not draw heroes in View World mode, just their icons
- the bottom of interface dawings are incorrect:
our version:
![image](https://user-images.githubusercontent.com/19829520/107649686-ac928d80-6cb8-11eb-8b87-6c32ccffa522.png)
the original game:
![image](https://user-images.githubusercontent.com/19829520/107649746-bc11d680-6cb8-11eb-96cc-614f4d38b333.png)

The reset of missing features like support of spells and letter we can do in a separate pull request.

vincent-grosbois commented:
Hello @ihhub 
please go ahead and either do the changes you feel are useful or let someone else do them, I don't think I will continue working on this issue.
This is probably the 10th back and forth we have on this PR, where each time a new objection is raised on the same initial codebase I've submitted
I think that for an open source project that concerns a 25 year old game, the focus should probably placed on trying to add missing/new features, rather than endlessly nitpicking which integers should be "cached" or not, which has absolutely 0 impact on the generated assembly and make the code less clear. I really don't think this is bringing anything to the final product, but this is wasting everybody's precious time

ihhub commented:
> Hello @ihhub
> please go ahead and either do the changes you feel are useful or let someone else do them, I don't think I will continue working on this issue.
> This is probably the 10th back and forth we have on this PR, where each time a new objection is raised on the same initial codebase I've submitted
> I think that for an open source project that concerns a 25 year old game, the focus should probably placed on trying to add missing/new features, rather than endlessly nitpicking which integers should be "cached" or not, which has absolutely 0 impact on the generated assembly and make the code less clear. I really don't think this is bringing anything to the final product, but this is wasting everybody's precious time

Hi @vincent-grosbois , sure. I will finish this pull request.

ihhub commented:
@vincent-grosbois , thank you so much for your huge work on this issue!

My apologies if I was too strict on performance optimizations as they are essential in such applications. Sadly fo us, the old code was written in a ba manner leading to have a mix of optimized and unoptimized solutions :(
